File: client/src/main/java/org/projectodd/openwhisk/InvokeOptions.java
package org.projectodd.openwhisk;

import org.projectodd.openwhisk.model.KeyValue;

import java.util.Map;

/**
 * Represents the options to apply to function invocations.
 */
public class InvokeOptions {
    private final String name;
    private boolean blocking;
    private boolean results;
    private int timeout = 30000;
    private KeyValue params = new KeyValue();

    public InvokeOptions(final String name) {
        this.name = name;
    }

    public String name() {
        return name;
    }

    public boolean blocking() {
        return blocking;
    }

    public InvokeOptions blocking(final boolean blocking) {
        this.blocking = blocking;
        return this;
    }

    public boolean results() {
        return results;
    }

    public InvokeOptions results(final boolean results) {
        this.results = results;
        return this;
    }

    public InvokeOptions parameter(String key, Object value) {
        params.put(key, value);
        return this;
    }

    public InvokeOptions parameters(Map<String, Object> values) {
//        params.addAll(values.entrySet().stream()
//              .map( e -> new KeyValue().key(e.getKey()).value(e.getValue()))
//              .collect(Collectors.toList()));
        params.putAll(values);
        return this;
    }

    KeyValue parameters() {
        return params;
    }

    public int timeout() {
        return timeout;
    }

    public InvokeOptions timeout(final int timeout) {
        this.timeout = timeout;
        return this;
    }
}

File: client/src/main/kotlin/org/projectodd/openwhisk/InvokeOptions.kt
package org.projectodd.openwhisk

import org.projectodd.openwhisk.model.KeyValue

/**
 * Represents the options to apply to function invocations.
 */
class InvokeOptions(private val name: String) {
    private var blocking: Boolean = false
    private var results: Boolean = false
    private var local: Boolean = false
    private var timeout = 30000
    private val params = KeyValue()

    fun name(): String {
        return name
    }

    fun blocking(): Boolean {
        return blocking
    }

    fun blocking(blocking: Boolean): InvokeOptions {
        this.blocking = blocking
        return this
    }

    fun local(): Boolean {
        return local
    }

    fun local(local: Boolean): InvokeOptions {
        this.local = local
        return this
    }

    fun results(): Boolean {
        return results
    }

    fun results(results: Boolean): InvokeOptions {
        this.results = results
        return this
    }

    fun parameter(key: String, value: Any): InvokeOptions {
        params[key] = value
        return this
    }

    fun parameters(values: Map<String, Any>): InvokeOptions {
        params.putAll(values)
        return this
    }

    internal fun parameters(): KeyValue {
        return params
    }

    fun timeout(): Int {
        return timeout
    }

    fun timeout(timeout: Int): InvokeOptions {
        this.timeout = timeout
        return this
    }
}
